lang/LangSource/SC_LanguageConfig.cpp

```diff
@@ -111,48 +114,64 @@ bool SC_LanguageConfig::removeExcludedDirectory(const Path& path)
 	return removePath(mExcludedDirectories, path);
 }
 
+void SC_LanguageConfig::setExcludeDefaultPaths(bool value)
+{
+	mExcludeDefaultPaths = value;
+}
+
+bool SC_LanguageConfig::getExcludeDefaultPaths() const
+{
+	return mExcludeDefaultPaths;
+}
+void SC_LanguageConfig::processPathList(const char* nodeName, YAML::Node &doc, void (*func)(const Path&)) {
+	const YAML::Node & items = doc[ nodeName ];
+	if (items && items.IsSequence()) {
+		std::string emptyString;
+		for (auto const & item : items ) {
+			const std::string& path = item.as<std::string>(emptyString);
+			if ( !path.empty() ) {
+				const Path& native_path = SC_Codecvt::utf8_str_to_path(path);
+				func(native_path);
+			}
+		}
+	}
+}
+
+void SC_LanguageConfig::processBool(const char* nodeName, YAML::Node &doc,  const std::function<void(bool)> &func) {
+	const YAML::Node & item = doc[ nodeName ];
+	if (item) {
+		try {
+			func(item.as<bool>());
+		} catch(...) {
+			postfl("Warning: Cannot parse config file entry \"%s\"\n", nodeName);
+		}
+	}
+}
+
 bool SC_LanguageConfig::readLibraryConfigYAML(const Path& fileName, bool standalone)
 {
 	freeLibraryConfig();
-	gLanguageConfig = new SC_LanguageConfig(standalone);
-
-	std::string emptyString;
 
 	using namespace YAML;
 	try {
 		bfs::ifstream fin(fileName);
 		Node doc = Load( fin );
 		if (doc) {
-			const Node & includePaths = doc[ INCLUDE_PATHS ];
-			if (includePaths && includePaths.IsSequence()) {
-				for (auto const & pathNode : includePaths ) {
-					const std::string& path = pathNode.as<std::string>(emptyString);
-					if ( !path.empty() ) {
-						const Path& native_path = SC_Codecvt::utf8_str_to_path(path);
-						gLanguageConfig->addIncludedDirectory(native_path);
-					}
-				}
-			}
-
-			const Node & excludePaths = doc[ EXCLUDE_PATHS ];
-			if (excludePaths && excludePaths.IsSequence()) {
-				for (auto const & pathNode : excludePaths ) {
-					const std::string& path = pathNode.as<std::string>(emptyString);
-					if ( !path.empty() ) {
-						const Path& native_path = SC_Codecvt::utf8_str_to_path(path);
-						gLanguageConfig->addExcludedDirectory(native_path);
-					}
-				}
-			}
-
-			const Node & inlineWarnings = doc[ POST_INLINE_WARNINGS ];
-			if (inlineWarnings) {
-				try {
-					gPostInlineWarnings = inlineWarnings.as<bool>();
-				} catch(...) {
-					postfl("WARNING: Cannot parse config file entry \"%s\"\n", POST_INLINE_WARNINGS);
-				}
-			}
+			processBool(POST_INLINE_WARNINGS, doc, [](bool b){
+				gPostInlineWarnings = b;
+			});
+			processBool(EXCLUDE_DEFAULT_PATHS, doc, [&standalone](bool b){
+				gLanguageConfig = new SC_LanguageConfig(b || standalone);
+			});
+			if (!gLanguageConfig)
+				gLanguageConfig = new SC_LanguageConfig(standalone);
```

Comment:
this logic was fine the way it was, with SC_LanguageConfig being created at the top of the function. You could use `SC_LanguageConfig::setExcludeDefaultPaths` here.

Reply:
Adding the default paths is done in the constructor of SC_LanguageConfig. You cannot call the constructor at the top of the function because you still haven't parsed the excludeDefaultPaths flag from the yaml file, and therefore you can't pass it to the constructor. This is why the object is created only after trying to get the excludeDefaultPaths flag from the yaml file. After having tried different approaches I came to prefer maintaining the constructor as it is now, where it adds the default paths in the constructor, I think the code is simpler that way.

Reply:
you could also assign a boolean and call the constructor later. more to the point, the design of this class should change so that a setter function is feasible. AFAICT there's no good reason not to allow that aside from the current implementation being what it is.

Reply:
> more to the point, the design of this class should change so that a setter function is feasible. AFAICT there's no good reason not to allow that aside from the current implementation being what it is.

My thoughts on this:

* The value of mDefaultClassLibraryDirectories is only relevant after reading from the YAML file and before compiling the class library (in PyrLexer.cpp passOne() ).  I don't think it is ever used again after that point. 
* Therefore mDefaultClassLibraryDirectories will have to be set very soon after the SC_LanguageConfig is created.
* Setting mDefaultClassLibraryDirectories in SC_LanguageConfig::setExcludeDefaultPaths(bool value) is ok, but actually it will not serve any purpose after the language is compiled, although that is also not a big issue, so I can implement it this way if you really prefer it.
* Given than mDefaultClassLibraryDirectories must be set after/while reading from the YAML file, might as well make this obvious, by initializing mDefaultClassLibraryDirectories in the constructor of SC_LanguageConfig which is called at that point.
* I think both options have pros and cons, so I'm OK with either option.

Reply:
Also, there is already a setter function, it just doesn't affect mDefaultClassLibraryDirectories for the reasons given above.  Unrelated to this issue, I will implement the primitive for the setter to be able to set it from sclang. That is only usefull when saving back to YAML.

Reply:
> Unrelated to this issue, I will implement the primitive for the setter to be able to set it from sclang. That is only usefull when saving back to YAML.

IMO that's reason enough to do it. If it isn't part of the LanguageConfig spec (hmm...), it should be: that all aspects of the configuration should be possible to set using language code, and written to disk.

IIRC the behavior is supposed to be:

- If you use sclang methods to alter the configuration and recompile the class library, the compiler should then use the in-memory (modified, new) image of the configuration immediately without re-reading the configuration from disk.
  - Use case: Test a configuration change in memory, without writing. If the change broke the environment, the original, working configuration is still on disk.

- If you modify the configuration outside of sclang (including IDE preferences), the only way to use the changes is to save them to disk and restart sclang. Recompiling the library doesn't work in that case.

So the setter would be useful also for recompiling the library.

(Again, confusion because nobody bothered to write down what is the expected behavior.)

Reply:
Ok, that is worth exploring. A slightly different approach could be:
* When recompiling the library the in-memory version is always used.
* On sclang startup it reads the language from disk.
* When changing the language config from the IDE, the IDE tells the language to re-read the configuration from disk, so as to pick up the changes when the language is recompiled. 
* Changes made to the config file manually or in other software will only be picked up after re-starting sclang.